docs/adr/adr-007-public-api.md

```diff
@@ -0,0 +1,194 @@
+implements all the possible APIs but with the variable implementations matching resource constraints of a type.
+
+## Design
+
+### Goals
+* Ergonomic. Simple, idiomatic and self-explanatory.
+* Module centric(modular). The API is not monolithic and segregated into different categorized and independent modules.
+* Unified. All the node types implement the same set of APIs. The difference is defined by different implementations of
+some modules to meet resource requirements of a type. Example: FullAvailability and LightAvailability
+* Embeddable. Simply constructable Node with library style API. Not an SDK/Framework which dictates users the way to build
+an app, but users are those who decide how they want to build the app using the API.
+* Language agnostic. It should be simple to implement similar module interfaces/traits in other languages over RPC clients
+
+### Details
+#### Services Deprecation
+The initial step is to deprecate and remove our services in code and stop relying on services as the notion. Instead, 
+the Module should be used, e.g. HeaderService -> HeaderModule.
+
+* We're organically moving towards the direction of modularized libraries. That is, our `share`, `header` and `state`
+packages are getting shaped as independent sovereign modules which now lack their own API definition. 
+* Consistency. Semantically, Modules are closer to what high level Celestia project goals and slogans.
+* Disassociate centralization. Services have always been associated with centralized infrastructure with older 
+monolithic services and newer distributed microservice architectures. 
+
+#### Modules Definition
+// TODO: Params and comment are yet to come
+##### Header
+```go
+type HeaderModule interface {
+    Head
+    Get
+    GetByHeight
+    GetRangeByHeight
+    Subscribe
+    SyncState
+    SyncWait
+}
+```
+
+##### Shares
+```go
+type SharesModule interface {
+	GetShare
+	Get
+	GetByNamespace
+	IsAvailable
+	Submit
+}
+```
+
+##### P2P/Net(work)
+```go
+type P2PModule/NetworkModule interface {
+    Info
+    Peers
+    PeerInfo
+
+    Connect
+    Disconnect
+    Connectedness
+    NATStatus
+
+    BlockAdd
+    BlockRm
+    BlockList
+
+    MutualAdd
+    MutualRm
+    MutualList
+
+    BandwidthStats
+    BandwidthPeer
+    BandwidthProtocol
+
+    ResourceState
+    ResourceLimit
+    ResourceSetLimit
+
+    PubSubPeers
+    PubSubScore
+
+    BitSwapStats
+    BitSwapLedger
+
+    DHTFindPeer
+}
+```
+
+##### Wallet
+```go
+type WalletModule interface {
+    New
+    Del
+    Has
+    List
+    Sign
+    Export
+    Import
+    // TODO: Mnemonics stuff?
+}
+```
+
+### NodeModule/MiscModule
+```go
+type NodeModule/MiscModule interface {
+    Type
+    Version
+
+    LogList
+    LogSet
+
+    // TODO: Shutdown/Stop?
+
+    // TODO: Auth
+
+}
+```
+
+##### DAS
+// TODO: Should we rebrand DASer to ShareSync?
+```go
+    State
+    Wait
```

Comment:
The more I think about it, the more I am convinced that `IsAvailable()` method belongs inside of `DASModule`. I think what would be a nice under-the-hood implementation of it is that `IsAvailable()` would check first in the [`ShareCache`](#180) whether shares available and if not, trigger a sampling routine on the given height. I think the `SharesModule` should deal more with share-specific methods, like the ones that already exist (`GetShare____` ).

Reply:
ShareCache is a CacheAvailabilty implementation of the Availavility interface which has IsAvailable method that Daser relies on. So Daser depends on this, not provides.

Let's arrange sync discussion about this.

Reply:
Wait is a method that allows users to block until Dasing is finished

Reply:
I would rename `Wait` then and have two methods

`IsAvailable()`
`WaitTilAvailable()` or something like that. 

`Wait` name makes me think that admin can pause `daser` routines.

Reply:
Update: 

@Wondertan and I had a call regarding where `IsAvailable` should live, and at the moment, let's just keep it inside `SharesModule` and if we find that people are using the `DASModule` more than expected, we can just provide the method there too as it will just use `Availability` implementation under the hood.